src/components/Header.tsx

```diff
@@ -0,0 +1,146 @@
+/* eslint-disable @typescript-eslint/indent */
+/* eslint-disable @typescript-eslint/no-unused-expressions */
+import { useEffect, useRef, useState } from 'react';
+import { Todo } from '../types/Todo';
+import { addTodo, updateTodo } from '../api/todos';
+
+interface HeaderProps {
+  todos: Todo[];
+  setEditingTodosId: (ids: number[]) => void;
+  setError: (error: string) => void;
+  editingTodosId: number[];
+  setTodos: (todos: Todo[]) => void;
+  setFilteredTodos: (todos: Todo[]) => void;
+  setIsActiveFilter: (isActive: boolean) => void;
+  filteredTodos: Todo[];
+}
+
+export function Header({
+  todos,
+  setEditingTodosId,
+  setTodos,
+  setFilteredTodos,
+  setError,
+  editingTodosId,
+}: HeaderProps) {
+  const [newTodoTitle, setNewTodoTitle] = useState('');
+  const [isAddingNewTodo, setIsAddingNewTodo] = useState(false);
+
+  const inputNewTodo = useRef<HTMLInputElement>(null);
+
+  function handleNewTodoTitle(e: React.ChangeEvent<HTMLInputElement>) {
+    setNewTodoTitle(e.target.value);
+  }
+
+  async function handleAddNewTodo(e: React.FormEvent<HTMLFormElement>) {
+    e.preventDefault();
+
+    if (!newTodoTitle.trim()) {
+      setError('Title should not be empty');
+      inputNewTodo.current?.focus();
+
+      return;
+    }
+
+    const newTodoIdAvailable = todos.length
+      ? todos[todos.length - 1].id + 1
+      : 1;
+
+    try {
+      setIsAddingNewTodo(true);
+      setEditingTodosId([...editingTodosId, newTodoIdAvailable]);
+
+      const newTodo = {
+        title: newTodoTitle.trim(),
+        completed: false,
+        id: newTodoIdAvailable,
+      };
+
+      setFilteredTodos([...todos, newTodo as Todo]);
+
+      const responseFromAddedTodo = await addTodo(newTodoTitle.trim());
+
+      setFilteredTodos([...todos].filter(t => t.id !== newTodoIdAvailable));
+      setTodos([...todos, responseFromAddedTodo]);
+      setFilteredTodos([...todos, responseFromAddedTodo]);
```

Comment:
This part of the code is a little confusing, you're manipolating the `filteredTodos` too much, making the rendered `todos` messy. Try imagining what you want to be rendered when you add a new `todo`, think of all scenarios, try adding a new `todo` when you have each filter on, and what you want to be rendered on the screen while adding the new `todo` in each case.